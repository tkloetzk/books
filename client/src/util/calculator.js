import get from 'lodash/get';
import forEach from 'lodash/forEach'
import clone from 'lodash/clone'

let meanGoodreadsVotes
let meanAmazonVotes
let totalMean

export default function sortBooklist(originalBooklist) {
  const booklist = clone(originalBooklist);
  calculateMeans(booklist);
  calculateAdjustedRating(booklist);
  booklist.sort((a, b) =>
    a.adjustedRating < b.adjustedRating
      ? 1
      : b.adjustedRating < a.adjustedRating
      ? -1
      : 0
  );
  return booklist;
}

function calculateMeans(booklist) {
  let goodreadsVotes = 0,
    amazonVotes = 0,
    total = 0;

  forEach(booklist, book => {
    goodreadsVotes += get(book, 'goodreadsAverageRating')
    amazonVotes += get(book, 'amazonAverageRating');
    total +=
      get(book, 'goodreadsAverageRating') +
      get(book, 'amazonAverageRating');
  })

  meanGoodreadsVotes = goodreadsVotes / booklist.length;
  meanAmazonVotes = amazonVotes / booklist.length;
  totalMean = total / 2 / booklist.length;

  return booklist;
}

function calculateAdjustedRating(booklist) {
  for (var i = 0; i < booklist.length; i++) {
    var book = booklist[i];
    var adjustedRating =
      (getAdjustedRating(
        get(book, 'goodreadsRatingsCount'),
        get(book, 'goodreadsAverageRating'),
        meanGoodreadsVotes,
        trimmean(getGoodreadsRatingsCountList(booklist))
      ) +
        getAdjustedRating(
          get(book, 'amazonRatingsCount'),
          get(book, 'amazonAverageRating'),
          meanAmazonVotes,
          trimmean(getAmazonRatingsCountList(booklist))
        )) /
      2;
    book.adjustedRating = adjustedRating;
  }
}

function getAdjustedRating(ratingsCount, averageRating, meanVote, minVotes) {
  //double rating = averageRating * 2;
  return (
    (ratingsCount / (ratingsCount + minVotes)) * averageRating +
    (minVotes / (ratingsCount + minVotes)) * meanVote
  );
}

function getGoodreadsRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.goodreadsRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function getAmazonRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.amazonRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function trimmean(votes) {
  votes.sort(function(a, b) {
    return a - b;
  });

  const removeAmount = (votes.length * 0.25) / 2;
  const roundDownNearedMultipleTwo = Math.floor(
    removeAmount >= 0
      ? (removeAmount / 2) * 2
      : ((removeAmount - 2 + 1) / 2) * 2
  );
  return getMean(
    votes.slice(
      roundDownNearedMultipleTwo,
      votes.length - roundDownNearedMultipleTwo
    )
  );
}

function getMean(array) {
  let sum = 0;
  for (let i = 0; i < array.length; i++) {
    sum += array[i];
  }
  return sum / array.length;
}
