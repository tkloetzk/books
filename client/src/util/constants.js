export const LOADING_STATUSES = {
  initial: 'initial',
  loading: 'loading',
  success: 'success',
  errored: 'errored',
};
