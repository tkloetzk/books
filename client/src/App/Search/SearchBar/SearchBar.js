import React, { Component } from 'react';
import { withStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import { getAmazonSingleBook } from '../../../store/amazon/amazonActions';
import { getGoodreadsBook } from '../../../store/goodreads/goodreadsActions';
import {
  saveCombinedBooks,
  saveModifiedBooks,
  addBookToBookshelf,
  updateBookOnBookshelf,
  clearBooks,
  insertModifiedBook,
} from '../../../store/bookshelf/bookshelfActions';
import { getGoogleBook } from '../../../store/google/googleActions';
import { connect } from 'react-redux';
import ReactTooltip from 'react-tooltip';
import CircularProgress from '@material-ui/core/CircularProgress';
import { LOADING_STATUSES } from '../../../util/constants';
import forEach from 'lodash/forEach';
import remove from 'lodash/remove';
import SaveIcon from '@material-ui/icons/Save';
import Fab from '@material-ui/core/Fab';
import map from 'lodash/map';
import assign from 'lodash/assign';
import Notification from '../../Notification/Notification';
import util from '../../../util/combineBooks';
import Tooltip from '../../Tooltip/Tooltip';

// TODO: This file is getting huge
const styles = {
  container: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'center',
    paddingBottom: '18px',
  },
  button: {
    marginTop: '15px',
    marginBottom: '8px',
    marginLeft: '15px',
    alignSelf: 'center',
  },
  buttonProgress: {
    color: 'green',
    position: 'relative',
    marginTop: 30,
    marginLeft: -59,
  },
  tooltipDiv: {
    zIndex: '1',
    position: 'absolute',
    width: 180,
    marginTop: -36,
  },
  fab: {
    alignSelf: 'center',
    marginLeft: '10px',
  },
};

export class SearchBar extends Component {
  state = {
    searchIsbns: [],
    duplicatedISBNs: [],
    multiline: '',
    loading: false,
    success: null,
    amazonBookLoading: LOADING_STATUSES.initial,
    goodreadsBookLoading: LOADING_STATUSES.initial,
    googleBookLoading: LOADING_STATUSES.initial,
  };

  componentDidUpdate(prevProps) {
    const {
      amazonBooks,
      goodreadsBooks,
      googleBooks,
      bookshelf,
      saveModifiedBooks,
      saveCombinedBooks,
      amazonBookErrored,
      goodreadsBooksErrored,
      googleBooksErrored,
      clearBooks,
      insertModifiedBook,
    } = this.props;
    const {
      searchIsbns,
      loading,
      amazonBookLoading,
      goodreadsBookLoading,
      googleBookLoading,
    } = this.state;

    if (searchIsbns.length) {
      if (
        amazonBooks.length === searchIsbns.length &&
        prevProps.amazonBooks.length !== amazonBooks.length
      ) {
        this.setState({ amazonBookLoading: LOADING_STATUSES.success });
      }
      if (
        goodreadsBooks.length === searchIsbns.length &&
        prevProps.goodreadsBooks.length !== goodreadsBooks.length
      ) {
        this.setState({ goodreadsBookLoading: LOADING_STATUSES.success });
      }
      if (
        googleBooks.length === searchIsbns.length &&
        prevProps.googleBooks.length !== googleBooks.length
      ) {
        this.setState({ googleBookLoading: LOADING_STATUSES.success });
      }
    }
    if (amazonBookErrored && amazonBookLoading !== LOADING_STATUSES.errored) {
      this.setState({
        loading: false,
        amazonBookLoading: LOADING_STATUSES.errored,
      });
    }
    if (
      goodreadsBooksErrored &&
      goodreadsBookLoading !== LOADING_STATUSES.errored
    ) {
      this.setState({
        loading: false,
        goodreadsBookLoading: LOADING_STATUSES.errored,
      });
    }
    if (googleBooksErrored && googleBookLoading !== LOADING_STATUSES.errored) {
      this.setState({
        loading: false,
        googleBookLoading: LOADING_STATUSES.errored,
      });
    }

    if (
      amazonBooks.length === searchIsbns.length &&
      googleBooks.length === searchIsbns.length &&
      goodreadsBooks.length === searchIsbns.length &&
      loading
    ) {
      // TODO: This all probably could be better
      const { combinedBooks, duplicates, duplicatedISBNs } = util.combineBooks(
        amazonBooks,
        googleBooks,
        goodreadsBooks,
        bookshelf
      );
      // TODO: What's this doing again?
      forEach([...duplicates, ...duplicatedISBNs], duplicate =>
        forEach([...combinedBooks], obj =>
          obj.isbn === duplicate.isbn ? remove(combinedBooks, obj) : null
        )
      );

      if (duplicates.length) {
        if (prevProps.modifiedBooklist.length) {
          insertModifiedBook(duplicates);
        } else {
          saveModifiedBooks(duplicates);
        }
      }
      if (combinedBooks.length) {
        saveCombinedBooks(combinedBooks);
      }
      clearBooks();
      this.setState({ success: true, loading: false, multiline: '' });
    }
  }
  handleChange = event => {
    this.setState({
      multiline: event.target.value,
    });
  };

  handleSearch = () => {
    const { getAmazonSingleBook, getGoogleBook, getGoodreadsBook } = this.props;
    const { multiline, loading } = this.state;
    const isbns = multiline.split(/[\n, ]/).filter(v => v !== '');
    if (!loading) {
      this.setState({
        success: false,
        loading: true,
        searchIsbns: isbns,
        goodreadsBookLoading: LOADING_STATUSES.loading,
        googleBookLoading: LOADING_STATUSES.loading,
        amazonBookLoading: LOADING_STATUSES.loading,
      });
    }

    Promise.all(
      forEach(isbns, isbn => {
        const formattedIsbn = isbn.replace(/[- ]/g, '');
        return [
          getAmazonSingleBook(formattedIsbn),
          getGoogleBook(formattedIsbn),
          getGoodreadsBook(formattedIsbn),
        ];
      })
    );
  };

  handleSave = () => {
    const {
      booklist,
      modifiedBooklist,
      addBookToBookshelf,
      updateBookOnBookshelf,
    } = this.props;
    if (modifiedBooklist.length) {
      Promise.all(
        forEach(modifiedBooklist, book => {
          const fields = map(book.differences, diff => {
            return { [diff.key]: diff.newValue };
          });
          return updateBookOnBookshelf(book._id, assign(...fields));
        })
      );
    }
    addBookToBookshelf(booklist).then(res => window.scrollTo(0, 0));
  };

  onClose = () => {
    this.setState({ duplicatedISBNs: [] });
  };
  render() {
    const { classes, booklist } = this.props;
    const {
      loading,
      duplicatedISBNs,
      amazonBookLoading,
      goodreadsBookLoading,
      googleBookLoading,
      multiline,
    } = this.state;
    const tooltipObj = [
      { label: 'Amazon', loading: amazonBookLoading },
      { label: 'Goodreads', loading: goodreadsBookLoading },
      { label: 'Google', loading: googleBookLoading },
    ];
    return (
      <form className={classes.container} noValidate autoComplete="off">
        <TextField
          id="outlined-full-width"
          multiline
          style={{ width: '700px' }}
          value={multiline}
          onChange={event => this.handleChange(event)}
          className={classes.textField}
          margin="normal"
          variant="outlined"
        />
        <span
          data-tip
          data-for="searchButtonWrapper"
          style={{ display: 'flex' }}
        >
          <Button
            variant="outlined"
            color="primary"
            className={classes.button}
            onClick={this.handleSearch}
            disabled={loading || !multiline.length}
          >
            Search
          </Button>
        </span>
        {booklist.length > 0 && (
          <Fab color="primary" aria-label="Save" className={classes.fab}>
            <SaveIcon onClick={this.handleSave} />
          </Fab>
        )}
        {loading && (
          <CircularProgress size={24} className={classes.buttonProgress} />
        )}
        <ReactTooltip
          id="searchButtonWrapper"
          place="right"
          effect="solid"
          getContent={() => <Tooltip content={tooltipObj} />}
        />
        <Notification
          open={false}
          //open={duplicatedISBNs.length ? true : false}
          handleClose={this.onClose}
          autoHideDuration={3500}
          message={`${duplicatedISBNs.join(
            ', '
          )} already shelved with no differences`}
          type={LOADING_STATUSES.info}
        />
      </form>
    );
  }
}

SearchBar.defaultProps = {
  classes: {},
  booklist: [],
  amazonBooks: [],
  googleBooks: [],
  goodreadsBooks: [],
  amazonBookErrored: false,
};

SearchBar.propTypes = {
  classes: PropTypes.object.isRequired,
  amazonBooks: PropTypes.array,
  booklist: PropTypes.array,
};

const mapStateToProps = state => {
  return {
    amazonBooks: state.amazon.books,
    amazonBookErrored: state.amazon.hasErrored,
    googleBooks: state.google.books,
    googleBooksErrored: state.google.hasErrored,
    goodreadsBooks: state.goodreads.books,
    goodreadsBooksErrored: state.goodreads.hasErrored,
    booklist: state.bookshelf.booklist,
    bookshelf: state.bookshelf.bookshelf,
    modifiedBooklist: state.bookshelf.modifiedBooklist,
  };
};
const mapDispatchToProps = {
  getAmazonSingleBook,
  getGoodreadsBook,
  getGoogleBook,
  saveCombinedBooks,
  saveModifiedBooks,
  addBookToBookshelf,
  updateBookOnBookshelf,
  clearBooks,
  insertModifiedBook,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withStyles(styles)(SearchBar));
