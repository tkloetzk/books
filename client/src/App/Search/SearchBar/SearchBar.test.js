import React from 'react';
import { shallow } from 'enzyme';
import { SearchBar } from './SearchBar';
import Button from '@material-ui/core/Button';
import { LOADING_STATUSES } from '../../../util/constants';
import cloneDeep from 'lodash/cloneDeep';

describe('SearchBar', () => {
  let props;
  let wrapper;
  let instance;

  beforeEach(() => {
    props = {
      classes: {},
      modifiedBooklist: [],
      clearBooks: jest.fn(),
      getAmazonSingleBook: jest.fn(),
      getGoogleBook: jest.fn(),
      getGoodreadsBook: jest.fn(),
      saveCombinedBooks: jest.fn(),
      amazonBookErrored: false,
      saveCombinedBooks: jest.fn(),
      insertModifiedBook: jest.fn(),
      saveModifiedBooks: jest.fn(),
    };
    wrapper = shallow(<SearchBar {...props} />);
    instance = wrapper.instance();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders', () => {
    expect(wrapper).toMatchSnapshot();
  });
  describe('search', () => {
    describe('button', () => {
      it('disabled if no text', () => {
        wrapper.setState({
          loading: false,
          multiline: '',
        });
        expect(
          wrapper
            .find(Button)
            .at(0)
            .props().disabled
        ).toBe(true);
      });
      it('disabled if loading', () => {
        wrapper.setState({
          loading: true,
          multiline: 'ISBN',
        });
        expect(
          wrapper
            .find(Button)
            .at(0)
            .props().disabled
        ).toBe(true);
      });
    });
    describe('bar', () => {
      it('handleChange updates state', () => {
        instance.handleChange({ target: { value: 'update state' } });
        expect(instance.state.multiline).toEqual('update state');
      });
    });
    describe('isbn array', () => {
      it('one long string with comma separator', () => {
        instance.state.multiline = '9780805835595,9780800731915,,9780375700002';
        instance.handleSearch();
        expect(instance.state.searchIsbns).toEqual([
          '9780805835595',
          '9780800731915',
          '9780375700002',
        ]);
      });
      it('one long string with space separator', () => {
        instance.state.multiline = '9780805835595 9780800731915  9780375700002';
        instance.handleSearch();
        expect(instance.state.searchIsbns).toEqual([
          '9780805835595',
          '9780800731915',
          '9780375700002',
        ]);
      });
      it('one long string with newline separator', () => {
        instance.state.multiline =
          '9780805835595\n9780800731915\n\n9780375700002';
        instance.handleSearch();
        expect(instance.state.searchIsbns).toEqual([
          '9780805835595',
          '9780800731915',
          '9780375700002',
        ]);
      });
    });
    describe('promise array', () => {
      it('function calls', () => {
        instance.state.multiline = '9780805835595,9780800731915,9780375700002';
        instance.handleSearch();

        expect(instance.props.getAmazonSingleBook).toHaveBeenCalledTimes(3);
        expect(instance.props.getGoogleBook).toHaveBeenCalledTimes(3);
        expect(instance.props.getGoodreadsBook).toHaveBeenCalledTimes(3);
      });
    });
  });
  describe('componentDidUpdate', () => {
    it('lifecycle of successful loading states', () => {
      const prevProps = {
        amazonBooks: [],
        goodreadsBooks: [],
        googleBooks: [],
      };
      expect(instance.state.amazonBookLoading).toEqual(
        LOADING_STATUSES.initial
      );
      expect(instance.state.goodreadsBookLoading).toEqual(
        LOADING_STATUSES.initial
      );
      expect(instance.state.googleBookLoading).toEqual(
        LOADING_STATUSES.initial
      );

      instance.state.multiline = '9780805835595,9780800731915';
      instance.handleSearch();

      instance.componentDidUpdate(prevProps);

      expect(instance.state.amazonBookLoading).toEqual(
        LOADING_STATUSES.loading
      );
      expect(instance.state.goodreadsBookLoading).toEqual(
        LOADING_STATUSES.loading
      );
      expect(instance.state.googleBookLoading).toEqual(
        LOADING_STATUSES.loading
      );

      wrapper.setProps({
        googleBooks: [{ title: 'title1' }, { title: 'title2' }],
        goodreadsBooks: [{ title: 'title1' }],
        amazonBooks: [{ title: 'title1' }],
      });
      expect(instance.state.googleBookLoading).toEqual(
        LOADING_STATUSES.success
      );
      expect(instance.state.amazonBookLoading).toEqual(
        LOADING_STATUSES.loading
      );
      expect(instance.state.goodreadsBookLoading).toEqual(
        LOADING_STATUSES.loading
      );

      wrapper.setProps({
        goodreadsBooks: [{ title: 'title1' }, { title: 'title2' }],
      });

      expect(instance.state.googleBookLoading).toEqual(
        LOADING_STATUSES.success
      );
      expect(instance.state.amazonBookLoading).toEqual(
        LOADING_STATUSES.loading
      );
      expect(instance.state.goodreadsBookLoading).toEqual(
        LOADING_STATUSES.success
      );

      wrapper.setProps({
        amazonBooks: [{ title: 'title1' }, { title: 'title2' }],
      });
      expect(instance.state.googleBookLoading).toEqual(
        LOADING_STATUSES.success
      );
      expect(instance.state.amazonBookLoading).toEqual(
        LOADING_STATUSES.success
      );
      expect(instance.state.goodreadsBookLoading).toEqual(
        LOADING_STATUSES.success
      );
    });
    it('errored loading states', () => {
      wrapper.setProps({ amazonBookErrored: true });
      instance.componentDidUpdate();
      expect(instance.state.amazonBookLoading).toEqual(
        LOADING_STATUSES.errored
      );
      wrapper.setProps({ goodreadsBooksErrored: true });
      instance.componentDidUpdate();
      expect(instance.state.goodreadsBookLoading).toEqual(
        LOADING_STATUSES.errored
      );
      wrapper.setProps({ googleBooksErrored: true });
      instance.componentDidUpdate();
      expect(instance.state.googleBookLoading).toEqual(
        LOADING_STATUSES.errored
      );
    });
    describe('books returned', () => {
      let amazonBooks;
      let googleBooks;
      let goodreadsBooks;
      let prevProps;
      let newBook;
      let modifiedBook;
      beforeEach(() => {
        amazonBooks = [
          {
            amazonAverageRating: 4.6,
            amazonRatingsCount: 113,
            price: '',
            isbn: '9780988995819',
          },
        ];
        goodreadsBooks = [
          {
            isbn: '9780988995819',
            goodreadsAverageRating: 4.22,
            goodreadsRatingsCount: 236,
          },
        ];
        googleBooks = [
          {
            title: 'Attachment Parenting',
            isbn: '9780988995819',
            subtitle: 'Instinctive Care for Your Baby and Young Child',
            description:
              'A complete guide to the concept of attachment parenting, which argues that parental responsiveness to a baby\'s needs leads to a well-adjusted child, offers tips on breastfeeding on demand, responding to a baby\'s cries, minimizing parent-child separation, and avoiding baby "gadgets." Original.',
            thumbnail:
              'http://books.google.com/books/content?id=daeq84DTC3IC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
            categories: ['Family & Relationships'],
          },
        ];
        prevProps = {
          amazonBooks: [],
          goodreadsBooks: [],
          googleBooks: [],
          modifiedBooklist: [],
        };
        newBook = [
          {
            amazonAverageRating: 4.6,
            amazonRatingsCount: 113,
            categories: ['Family & Relationships'],
            description:
              'A complete guide to the concept of attachment parenting, which argues that parental responsiveness to a baby\'s needs leads to a well-adjusted child, offers tips on breastfeeding on demand, responding to a baby\'s cries, minimizing parent-child separation, and avoiding baby "gadgets." Original.',
            goodreadsAverageRating: 4.22,
            goodreadsRatingsCount: 236,
            isbn: '9780988995819',
            price: '',
            subtitle: 'Instinctive Care for Your Baby and Young Child',
            thumbnail:
              'http://books.google.com/books/content?id=daeq84DTC3IC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
            title: 'Attachment Parenting',
          },
        ];
        modifiedBook = [
          {
            _id: undefined,
            amazonAverageRating: 2,
            amazonRatingsCount: 7,
            categories: ['Family & Relationships'],
            description:
              'A complete guide to the concept of attachment parenting, which argues that parental responsiveness to a baby\'s needs leads to a well-adjusted child, offers tips on breastfeeding on demand, responding to a baby\'s cries, minimizing parent-child separation, and avoiding baby "gadgets." Original.',
            differences: [
              {
                currentValue: 4.6,
                key: 'amazonAverageRating',
                newValue: 2,
              },
              {
                currentValue: 113,
                key: 'amazonRatingsCount',
                newValue: 7,
              },
            ],
            goodreadsAverageRating: 4.22,
            goodreadsRatingsCount: 236,
            isbn: '9780988995819',
            price: '',
            subtitle: 'Instinctive Care for Your Baby and Young Child',
            thumbnail:
              'http://books.google.com/books/content?id=daeq84DTC3IC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
            title: 'Attachment Parenting',
          },
        ];
      });
      it('new book', () => {
        props = Object.assign({}, props, {
          googleBooks,
          goodreadsBooks,
          amazonBooks,
        });

        wrapper = shallow(<SearchBar {...props} />);
        instance = wrapper.instance();
        instance.state.loading = true;
        instance.state.searchIsbns = ['9780988995819'];

        instance.componentDidUpdate(prevProps);
        expect(instance.props.insertModifiedBook).not.toHaveBeenCalled();
        expect(instance.props.saveModifiedBooks).not.toHaveBeenCalled();
        expect(instance.props.saveCombinedBooks).toBeCalledWith(newBook);
      });
      it('modified book', () => {
        amazonBooks[0].amazonAverageRating = 2;
        amazonBooks[0].amazonRatingsCount = 7;
        props = Object.assign({}, props, {
          googleBooks,
          goodreadsBooks,
          amazonBooks,
          bookshelf: newBook,
        });
        wrapper = shallow(<SearchBar {...props} />);
        instance = wrapper.instance();
        instance.state.loading = true;
        instance.state.searchIsbns = ['9780988995819'];

        instance.componentDidUpdate(prevProps);

        expect(instance.props.insertModifiedBook).not.toHaveBeenCalled();
        expect(instance.props.saveModifiedBooks).toHaveBeenCalledWith(
          modifiedBook
        );
        expect(instance.props.saveCombinedBooks).not.toHaveBeenCalled();
      });
      it('exisiting book with no differences', () => {
        // TODO: get this implemented
      });
      it('new books with modified book', () => {
        amazonBooks[0].amazonAverageRating = 2;
        amazonBooks[0].amazonRatingsCount = 7;
        amazonBooks.push({
          amazonAverageRating: 4.6,
          amazonRatingsCount: 932,
          price: '',
          isbn: '9780553807912',
        });
        googleBooks.push({
          amazonAverageRating: 4.6,
          amazonRatingsCount: 932,
          price: '',
          isbn: '9780553807912',
        });
        goodreadsBooks.push({
          isbn: '9780553807912',
          goodreadsAverageRating: 4.18,
          goodreadsRatingsCount: 11600,
        });
        props = Object.assign({}, props, {
          googleBooks,
          goodreadsBooks,
          amazonBooks,
          bookshelf: newBook,
        });
        wrapper = shallow(<SearchBar {...props} />);
        instance = wrapper.instance();
        instance.state.loading = true;
        instance.state.searchIsbns = ['9780988995819', '9780553807912'];

        instance.componentDidUpdate(prevProps);

        expect(instance.props.insertModifiedBook).not.toHaveBeenCalled();
        expect(instance.props.saveModifiedBooks).toHaveBeenCalledWith(
          modifiedBook
        );
        expect(instance.props.saveCombinedBooks).toHaveBeenCalledWith([
          {
            amazonAverageRating: 4.6,
            amazonRatingsCount: 932,
            goodreadsAverageRating: 4.18,
            goodreadsRatingsCount: 11600,
            isbn: '9780553807912',
            price: '',
          },
        ]);
      });
      it('modified book with previous modified book', () => {});
      it('new books, modified books, exisiting with no differences', () => {
        // TODO: get this implemented
      });
    });
  });
});
