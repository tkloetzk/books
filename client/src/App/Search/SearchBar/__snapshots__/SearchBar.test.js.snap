// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Saving Fab save shows save button if there are books in booklist and not in modifiedbooklist 1`] = `
<button
  aria-label="Save"
  className="MuiButtonBase-root-66 MuiFab-root-119 MuiFab-primary-121"
  disabled={false}
  id="SaveIcon"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiFab-label-120"
  >
    <pure(SaveIcon)>
      <SaveIcon>
        <WithStyles(SvgIcon)>
          <SvgIcon
            classes={
              Object {
                "colorAction": "MuiSvgIcon-colorAction-108",
                "colorDisabled": "MuiSvgIcon-colorDisabled-110",
                "colorError": "MuiSvgIcon-colorError-109",
                "colorPrimary": "MuiSvgIcon-colorPrimary-106",
                "colorSecondary": "MuiSvgIcon-colorSecondary-107",
                "fontSizeInherit": "MuiSvgIcon-fontSizeInherit-111",
                "fontSizeLarge": "MuiSvgIcon-fontSizeLarge-113",
                "fontSizeSmall": "MuiSvgIcon-fontSizeSmall-112",
                "root": "MuiSvgIcon-root-105",
              }
            }
            color="inherit"
            component="svg"
            fontSize="default"
            viewBox="0 0 24 24"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-105"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
          </SvgIcon>
        </WithStyles(SvgIcon)>
      </SaveIcon>
    </pure(SaveIcon)>
  </span>
  <NoSsr
    defer={false}
    fallback={null}
  >
    <WithStyles(TouchRipple)
      center={false}
      innerRef={[Function]}
    >
      <TouchRipple
        center={false}
        classes={
          Object {
            "child": "MuiTouchRipple-child-133",
            "childLeaving": "MuiTouchRipple-childLeaving-134",
            "childPulsate": "MuiTouchRipple-childPulsate-135",
            "ripple": "MuiTouchRipple-ripple-130",
            "ripplePulsate": "MuiTouchRipple-ripplePulsate-132",
            "rippleVisible": "MuiTouchRipple-rippleVisible-131",
            "root": "MuiTouchRipple-root-129",
          }
        }
      >
        <TransitionGroup
          childFactory={[Function]}
          className="MuiTouchRipple-root-129"
          component="span"
          enter={true}
          exit={true}
        >
          <span
            className="MuiTouchRipple-root-129"
          />
        </TransitionGroup>
      </TouchRipple>
    </WithStyles(TouchRipple)>
  </NoSsr>
</button>
`;

exports[`SearchBar Saving Fab save shows save button if there are no books in booklist but in modifiedBooklist 1`] = `
<button
  aria-label="Save"
  className="MuiButtonBase-root-66 MuiFab-root-119 MuiFab-primary-121"
  disabled={false}
  id="SaveIcon"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiFab-label-120"
  >
    <pure(SaveIcon)>
      <SaveIcon>
        <WithStyles(SvgIcon)>
          <SvgIcon
            classes={
              Object {
                "colorAction": "MuiSvgIcon-colorAction-108",
                "colorDisabled": "MuiSvgIcon-colorDisabled-110",
                "colorError": "MuiSvgIcon-colorError-109",
                "colorPrimary": "MuiSvgIcon-colorPrimary-106",
                "colorSecondary": "MuiSvgIcon-colorSecondary-107",
                "fontSizeInherit": "MuiSvgIcon-fontSizeInherit-111",
                "fontSizeLarge": "MuiSvgIcon-fontSizeLarge-113",
                "fontSizeSmall": "MuiSvgIcon-fontSizeSmall-112",
                "root": "MuiSvgIcon-root-105",
              }
            }
            color="inherit"
            component="svg"
            fontSize="default"
            viewBox="0 0 24 24"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-105"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
          </SvgIcon>
        </WithStyles(SvgIcon)>
      </SaveIcon>
    </pure(SaveIcon)>
  </span>
  <NoSsr
    defer={false}
    fallback={null}
  >
    <WithStyles(TouchRipple)
      center={false}
      innerRef={[Function]}
    >
      <TouchRipple
        center={false}
        classes={
          Object {
            "child": "MuiTouchRipple-child-133",
            "childLeaving": "MuiTouchRipple-childLeaving-134",
            "childPulsate": "MuiTouchRipple-childPulsate-135",
            "ripple": "MuiTouchRipple-ripple-130",
            "ripplePulsate": "MuiTouchRipple-ripplePulsate-132",
            "rippleVisible": "MuiTouchRipple-rippleVisible-131",
            "root": "MuiTouchRipple-root-129",
          }
        }
      >
        <TransitionGroup
          childFactory={[Function]}
          className="MuiTouchRipple-root-129"
          component="span"
          enter={true}
          exit={true}
        >
          <span
            className="MuiTouchRipple-root-129"
          />
        </TransitionGroup>
      </TouchRipple>
    </WithStyles(TouchRipple)>
  </NoSsr>
</button>
`;

exports[`SearchBar renders 1`] = `
<form
  autoComplete="off"
  noValidate={true}
>
  <TextField
    id="outlined-full-width"
    margin="normal"
    multiline={true}
    onChange={[Function]}
    required={false}
    select={false}
    style={
      Object {
        "width": "700px",
      }
    }
    value=""
    variant="outlined"
  />
  <span
    data-for="searchButtonWrapper"
    data-tip={true}
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <WithStyles(Button)
      color="primary"
      disabled={true}
      onClick={[Function]}
      variant="outlined"
    >
      Search
    </WithStyles(Button)>
  </span>
  <ReactTooltip
    effect="solid"
    getContent={[Function]}
    id="searchButtonWrapper"
    insecure={true}
    place="right"
    resizeHide={true}
    wrapper="div"
  />
  <WithStyles(Notification)
    autoHideDuration={3500}
    handleClose={[Function]}
    message=" already shelved with no differences"
    open={false}
  />
</form>
`;
